src/main/kotlin/com/mdrobnak/lalrpop/inspections/WrongInferredTypeInspection.kt

```diff
@@ -1,36 +1,77 @@
+import org.rust.lang.core.resolve.ImplLookup
+import org.rust.lang.core.resolve.knownItems
+import org.rust.lang.core.types.infer.RsInferenceContext
 
 /**
  * Reports 2 kinds of problems on a nonterminal
  * 1. The inferred type is different from the type declared explicitly on the nonterminal
  * 2. There are 2 different alternatives, without action code, where the inferred types are different
  */
 object WrongInferredTypeInspection : LocalInspectionTool() {
-    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+    override fun buildVisitor(
+        holder: ProblemsHolder,
+        isOnTheFly: Boolean,
+        session: LocalInspectionToolSession
+    ): PsiElementVisitor {
         return object : LpVisitor() {
             override fun visitNonterminal(nonterminal: LpNonterminal) {
+                if (nonterminal.nonterminalName.nonterminalParams != null) return
+
                 val explicitType = nonterminal.typeRef?.resolveType(listOf())
 
+                if (explicitType == "()") return
+
                 var seenType = explicitType
 
                 nonterminal.alternatives.alternativeList.filter { it.action == null }.forEach { alternative ->
                     val alternativeType = alternative.resolveType(listOf())
 
-                    if (seenType != null && seenType != alternativeType) { // TODO: maybe get the rust plugin to do this comparison because it may be very error-prone by just comparing strings
-                        holder.registerProblem(alternative, "Resolved type of alternative is `$alternativeType`, while expected type `$seenType`")
-                    }
-
                     if (seenType == null) {
                         seenType = alternativeType
+                        return
+                    }
+
+                    if (!sameTypes(
+                            nonterminal.project,
+                            nonterminal.containingFile.importCode,
+                            seenType!!,
+                            alternativeType
+                        )
+                    ) {
+                        holder.registerProblem(
+                            alternative,
+                            "Resolved type of alternative is `$alternativeType`, while expected type `$seenType`"
+                        )
                     }
                 }
             }
         }
     }
+
+    fun sameTypes(project: Project, importCode: String, type1: String, type2: String): Boolean {
+        val file = PsiFileFactory.getInstance(project)
+            .createFileFromText(RsLanguage, "$importCode\ntype T1 = $type1;\ntype T2 = $type2;")
+        val aliases = file.childrenOfType<RsTypeAlias>();
+
+        if (aliases.size != 2) return false
+
+        return RsInferenceContext(project, ImplLookup.relativeTo(aliases[0]), aliases[0].knownItems).canCombineTypes(
+            aliases[0].declaredType,
+            aliases[1].declaredType
```

Comment:
For some reason the declared types in the following case are `TyUnknown`:

In leafbuild, after modifying the type of `Num` from inferred to an (incorrect) explicit, no errors are shown:
![image](https://user-images.githubusercontent.com/4417660/101258583-67b70d00-36f1-11eb-8c64-66264c24f313.png)

Reply:
Hmm yes forgot to do that `attachInjectedCodeToCrate` or something similar. Went with:

```kotlin
    fun sameTypes(project: Project, lalrpopFile: PsiFile, importCode: String, type1: String, type2: String): Boolean {
        val file = PsiFileFactory.getInstance(project)
            .createFileFromText(RsLanguage, "$importCode\ntype T1 = $type1;\ntype T2 = $type2;")
        val aliases = file.childrenOfType<RsTypeAlias>();

        if (aliases.size != 2) return false

        val moduleDefinition = findModuleDefinition(project, lalrpopFile) ?: return false

        for (child in file.childrenOfType<RsExpandedElement>()) {
            child.setContext(moduleDefinition)
        }

        val ctx = ImplLookup.relativeTo(aliases[0]).ctx
        val ty1 = ctx.fullyResolve(aliases[0].declaredType)
        val ty2 = ctx.fullyResolve(aliases[1].declaredType)

        println("First: $type1, second: $type2, aliases: ${aliases[0].text}, ${aliases[1].text}, ty1 = ${ty1.printToString()}, ty2 = ${ty2.printToString()}")

        return ctx.canCombineTypes(ty1, ty2)
    }
```

And now I only know it cannot resolve crate types, while it can resolve types in stdlib. Will play with this a little more